vip-helpers/vip-newrelic.php

```diff
@@ -0,0 +1,125 @@
+		newrelic_add_custom_parameter( 'HTTP_USER_AGENT', isset( $_SERVER['HTTP_USER_AGENT'] ) ? $_SERVER['HTTP_USER_AGENT'] : '' );
+		newrelic_add_custom_parameter( 'HTTPS', is_ssl() );
+	}
+}
+
+/**
+ * Name cron correctly in New Relic and do not count it as part of the Apdex score.
+ *
+ * We don't want to count cron as part of the apdex because it is not a user facing function and if cron tasks are slow it doesn't imply that the site's performance is impacted. Without removing these, long running cron tasks could flag the site as having performance problems, which would cause false positives in the monitoring.
+ */
+function wpcom_vip_cron_for_newrelic() {
+	if ( wp_doing_cron()
+		&& function_exists( 'newrelic_ignore_apdex' )
+		&& function_exists( 'newrelic_add_custom_parameter' )
+		&& function_exists( 'newrelic_name_transaction' ) ) {
+		newrelic_name_transaction( 'wp-cron' );
+		// N.B. VIP Go sites execute cron events through WP CLI, so the event
+		// can be determined in the New Relic 'wp-cli-cmd-args' custom parameter
+		newrelic_ignore_apdex();
+	}
+}
+
+/**
+ * Name wp-cli correctly in New Relic and do not count it as part of the Apdex score
+ *
+ * We don't want to count ongoing WP-CLI requests as part of the apdex because it is not a user facing function and if a WP-CLI request is slow it doesn't imply that the site's performance is impacted. Without removing these WP-CLI requests from the apdex calculation it could flag the site as having performance problems, which would cause false positives in the monitoring.
+ */
+function wpcom_vip_wpcli_for_newrelic() {
+	if ( defined( 'WP_CLI' )
+	     && WP_CLI
+	     && ! wp_doing_cron()  // Cron is going to be run via WP-CLI in the near term. We want to keep Cron and WP-CLI separated for better monitoring so we're not going to flag WP_CLI requests that are actually cron requests as WP-CLI.
+	     && function_exists( 'newrelic_ignore_apdex' )
+	     && function_exists( 'newrelic_add_custom_parameter' )
+	     && function_exists( 'newrelic_name_transaction' ) ) {
+
+		newrelic_name_transaction( 'wp-cli' );
+
+		$cmd = 'wp ' . implode( ' ', \WP_CLI::get_runner()->arguments );
+		// e.g. `wp option get siteurl`
+		newrelic_add_custom_parameter( 'wp-cli-cmd', $cmd );
+
+		$assoc_args = \WP_CLI::get_runner()->assoc_args;
+		$config = \WP_CLI::get_runner()->config;
+		// Empty values clutter the result
+		foreach ( $config as $config_arg => $config_value ) {
+			if ( empty( $config_value ) ) {
+				unset( $config[$config_arg] );
+			}
+		}
+		$all_config = array_merge( $assoc_args, $config );
+		// Key sorting normalises the resultant string
+		ksort( $all_config );
+		// e.g. `allow-root=1&color=auto&format=json&skip-plugins=1` which is
+		// equivalent to `--allow-root --color=auto --format=json --skip-plugins`
+		// We have this format because building a query string is easy and understandable
+		// (and a hack) whereas rebuilding a WP CLI style command parameters is harder
```

Comment:
Would we prefer to manually rebuild in the style of WP CLI, e.g. `--allow-root --color=auto --format=json --skip-plugins`, or are we happy enough with the query string style representation?

Reply:
I'm not sure. Maybe we should start with this and see if further effort is warranted?

Reply:
I think this is good enough for now

Reply:
We need to be careful here since wp-cli commands may pass secret args like passwords. We should not be logging those.

I think just the command and subcommand are probably good enough for now. If we want to include other args, we need to whitelist them (but I don't think that's scalable.